system/daemon.go

```diff
@@ -82,176 +102,246 @@ func Supported(filename string) bool {
-
-func (sys *Daemon) Stop(name string) (err error) {
-	return nil
+	return tr.Run()
 }
 
-func (sys *Daemon) Restart(name string) (err error) {
-	if err = sys.Stop(name); err != nil {
+func (sys *Daemon) Restart(names ...string) (err error) {
+	var tr *transaction
+	if tr, err = sys.newTransaction(restart, names); err != nil {
 		return
 	}
-	return sys.Start(name)
+	return tr.Run()
 }
 
-func (sys *Daemon) Reload(name string) (err error) {
-	var u *Unit
-	if u, err = sys.Get(name); err != nil {
+func (sys *Daemon) Reload(names ...string) (err error) {
+	var tr *transaction
+	if tr, err = sys.newTransaction(reload, names); err != nil {
 		return
 	}
+	return tr.Run()
+}
 
-	if reloader, ok := u.Interface.(unit.Reloader); ok {
-		return reloader.Reload()
-	}
+func (sys *Daemon) newTransaction(typ jobType, names []string) (tr *transaction, err error) {
+	sys.mutex.Lock()
+	defer sys.mutex.Unlock()
+
+	tr = newTransaction()
+
+	for _, name := range names {
+		var dep *Unit
+		if dep, err = sys.Get(name); err != nil {
+			return nil, err
+		}
 
-	return ErrNoReload
+		if err = tr.add(typ, dep, nil, true, true); err != nil {
+			return nil, err
+		}
+	}
+	return
 }
 
 // TODO
-func (sys *Daemon) Enable(name string) (err error) {
-	var u *Unit
-	if u, err = sys.Get(name); err != nil {
-		return
-	}
-	u.Log.Println("enable")
-	return ErrNotImplemented
+func (sys *Daemon) Enable(names ...string) (err error) {
+	return sys.getAndExecute(names, func(u *Unit, gerr error) error {
+		if gerr != nil {
+			return gerr
+		}
+
+		return u.Enable()
+	})
 }
 
 // TODO
-func (sys *Daemon) Disable(name string) (err error) {
-	var u *Unit
-	if u, err = sys.Get(name); err != nil {
-		return
-	}
-	u.Log.Println("disable")
-	return ErrNotImplemented
+func (sys *Daemon) Disable(names ...string) (err error) {
+	return sys.getAndExecute(names, func(u *Unit, gerr error) error {
+		if gerr != nil {
+			return gerr
+		}
+
+		return u.Disable()
+	})
 }
 
-// IsEnabled returns enable state of the unit held in-memory under specified name
-// If error is returned, it is going to be ErrNotFound
-func (sys *Daemon) IsEnabled(name string) (st unit.Enable, err error) {
-	//var u *Unit
-	//if u, err = sys.Unit(name); err == nil && sys.Enabled[u] {
-	//st = unit.Enabled
-	//}
-	return unit.Enabled, ErrNotImplemented
+func (sys *Daemon) getAndExecute(names []string, fn func(*Unit, error) error) (err error) {
+	for _, name := range names {
+		if err = fn(sys.Get(name)); err != nil {
+			return
+		}
+	}
+	return
 }
 
-// IsActive returns activation state of the unit held in-memory under specified name
-// If error is returned, it is going to be ErrNotFound
-func (sys *Daemon) IsActive(name string) (st unit.Activation, err error) {
-	var u *Unit
-	if u, err = sys.Get(name); err == nil {
-		st = u.Active()
+// Units returns a slice of all units created
+func (sys *Daemon) Units() (units []*Unit) {
+	unitSet := map[*Unit]struct{}{}
+	for _, u := range sys.units {
+		unitSet[u] = struct{}{}
```

Comment:
are there duplicate units in `sys.units`?  or is this because there can be multiple names for a single unit?